src/Processors/Executors/PullingPipelineExecutor.h

```diff
@@ -0,0 +1,54 @@
+#pragma once
+#include <memory>
+
+namespace DB
+{
+
+class QueryPipeline;
+class Block;
+class Chunk;
+class LazyOutputFormat;
+struct BlockStreamProfileInfo;
+
+/// Pulling executor for QueryPipeline.
+/// Typical usage is:
+///
+/// PullingPipelineExecutor executor(query_pipeline);
+/// while (executor.pull(chunk))
+///     ... process chunk ...
+class PullingPipelineExecutor
+{
+public:
+    explicit PullingPipelineExecutor(QueryPipeline & pipeline_);
+    ~PullingPipelineExecutor();
+
+    /// Methods return false if query is finished.
+    /// If milliseconds > 0, returns empty object and `true` after timeout exceeded.
+    /// You can use any pull method.
+    bool pull(Chunk & chunk, uint64_t milliseconds = 0);
```

Comment:
Why `uint64_t ` but not `size_t`?

Reply:
Actually, I don't know :)
Probably, size_t is better for array indexes, and uint64_t is better for time.